src/debugging-support-in-rustc.md

```diff
@@ -0,0 +1,259 @@
+# Debugging support in the Rust compiler
+
+This document explains the state of debugging tools support in the Rust compiler (rustc).
+The document gives an overview of debugging tools like GDB, LLDB etc. and infrastrcture 
+around Rust compiler to debug Rust code. If you want to learn how to debug the Rust compiler
+itself, then you must see [Debugging the Compiler] page.
+
+The material gathered from [Tom Tromey discusses debugging support in rustc].
+
+## Preliminaries
+
+### Debuggers
+
+TODO: A brief overview of what debuggers are and why we need support of debuggers in 
+Rust compiler.
+
+### DWARF
+
+According to the [DWARF] standard website 
+
+> DWARF is a debugging file format used by many compilers and debuggers to support source level 
+debugging. It addresses the requirements of a number of procedural languages, 
+such as C, C++, and Fortran, and is designed to be extensible to other languages. 
+DWARF is architecture independent and applicable to any processor or operating system. 
+It is widely used on Unix, Linux and other operating systems, 
+as well as in stand-alone environments.
+
+DWARF reader is a program that consumes the DWARF format and creates debugger compatible output.
+This program may live in the compiler itself.  DWARF uses a data structure called 
+Debugging Information Entry (DIE) which stores the information as "tags" to denote functions, 
+variables etc., e.g., `DW_TAG_variable`, `DW_TAG_pointer_type`, `DW_TAG_subprogram` etc. 
+You can also invent your own tags and attributes.
+
+## Supported debuggers
+
+### GDB
+
+We have our own fork of GDB - [https://github.com/rust-dev-tools/gdb]
+
+#### Rust expression parser
+
+To be able to show debug output we need expression parser. 
+This expression parser is written in [Bison] and is only a subset of Rust expressions. 
+This means that this parser can parse only a subset of Rust expressions. 
+Parser was written from scratch and has no relation to any other parser. 
+For example, this parser is not related to Rustc's parser.
+
+GDB has Rust like value and type output. It can print values and types in a way 
+that look like Rust syntax in the output. Or when you print a type as [ptype] in GDB, 
+it also looks like Rust source code. Checkout the documentation in the [manual for GDB/Rust]. 
+The [manual for GDB/Rust] lists some limitations as well.
+
+#### Parser extensions
+
+Expression parser has a couple of extensions [TODO: Where are the extensions???] in it 
+to facilitate features that you cannot do with Rust [TODO: CHECK: Find out what cannot be done?].
```

Comment:
Not sure what cannot be done but this was mentioned by @tromey in the video. Perhaps he did explain later which I have noted in this document below, but I still am not sure if there is something more to this.

Reply:
gdb's Rust extensions and limitations are documented in the gdb manual: https://sourceware.org/gdb/onlinedocs/gdb/Rust.html -- however, this neglects to mention that gdb convenience variables and registers follow the gdb `$` convention, and that the Rust parser implements the gdb `@` extension.

Reply:
> however, this neglects to mention that gdb convenience variables and registers follow the gdb $ convention, and that the Rust parser implements the gdb @ extension.

@tromey do you think we should mention this part in the GDB-Rust document rather than this document so there is no duplication etc.?